### PR TITLE
Attempt to resolve 1.21.70 inventory issues

### PR DESCRIPTION
**Pull request by onebeastchris**
this is yet another attempt to resolve issues related to bedrock clients rejecting inventories instead of opening them.

This PR attempts to address it using multiple methods:
- Queuing opening of virtual containers using a NetworkStackLatencyPacket
- Ensuring current inventories and inventory translators don't get out of sync
- In the rare case that the client still rejects opening an inventory, the BedrockContainerCloseTranslator will attempt to re-open the inventory using yet another NetworkStackLatencyPacket hack

This PR further has quite aggressive debug logging related to inventories. If you, the person reading this, is still experiencing issues with inventories *even with this PR*, do the following:
- Enable `debug-mode` in the Geyser config.yml, and restart the server
- Join the server and play for a while until you (or any user) runs into this issue again
- Write a comment here with a geyser dump, and a full server log showing (ideally uploaded via mclo.gs)

You can download the PR builds here:
-  [Geyser-Bungeecord.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/bungeecord)
- [Geyser-Fabric.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/fabric)
- [Geyser-NeoForge.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/neoforge)
- [Geyser-Spigot.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/spigot)
- [Geyser-Standalone.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/standalone)
- [Geyser-Velocity.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/velocity)
- [Geyser-ViaProxy.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.5443/builds/latest/downloads/viaproxy)

In my extensive testing, this PR seems to work fine. It is hacky, and ugly, but that is the true nature of Minecraft Bedrock edition, and that cannot change. ~~This PR *initially* attempted to work around that by re-using virtual inventories. That bit of code is currently commented out as it's not clear whether there would be a tangible benefit to implementing it. ~~ hahaha chris very funny. Some plugins send multiple ClientboundContainerOpen packets with the same inventory - not reusing inventories would be very annoying.

fixes https://github.com/GeyserMC/Geyser/issues/5440
fixes https://github.com/GeyserMC/Geyser/issues/5439
fixes https://github.com/GeyserMC/Geyser/issues/5437

### CODE REVIEW

File: core/src/main/java/org/geysermc/geyser/util/InventoryUtils.java [hunk 1]
```diff
@@ -140,7 +181,8 @@ public static void closeInventory(GeyserSession session, int javaId, boolean con
             }
             session.getBundleCache().onInventoryClose(inventory);
         }
-        session.setInventoryTranslator(InventoryTranslator.PLAYER_INVENTORY_TRANSLATOR);
+
+        GeyserImpl.getInstance().getLogger().sessionDebugLog(session, "Closed inventory: " + inventory.getJavaId() + " is waiting on confirm?" + session.isClosingInventory());
```
Comment (onebeastchris) [inline core/src/main/java/org/geysermc/geyser/util/InventoryUtils.java:185]:
```suggestion
        GeyserImpl.getInstance().getLogger().sessionDebugLog(session, "Closed inventory: " + (inventory != null ? inventory.getJavaId() : "null") + " is waiting on confirm?" + session.isClosingInventory());
```

**Comment by jackh54** [comment]
Hey, I tested this on my server—unfortunately, it does keep inventories open, but there are some issues when interacting with GUIs. For example, I use ShopGUIPlus, and none of the shop buttons respond or open the shop interfaces properly. It also leaves a ghost chest floating above the player (see attached video).

https://github.com/user-attachments/assets/3c1b5ef5-04ea-43a9-9745-f71526659f08

**Comment by eclipseisoffline** [comment]
@jackh54 as is said in the PR description, could you please do the following to help us debug the issue:

> This PR further has quite aggressive debug logging related to inventories. If you, the person reading this, is still experiencing issues with inventories even with this PR, do the following:
> 
> - Enable `debug-mode` in the Geyser config.yml, and restart the server
> - Join the server and play for a while until you (or any user) runs into this issue again
> - Write a comment here with a geyser dump, and a full server log showing (ideally uploaded via mclo.gs)

**Comment by Uniicz** [comment]
Bedrock players can't open GUI from other plugins. In this case, they can't use /shop and select a shop. (EconomyShopGUI)

DUMP File:
https://dump.geysermc.org/qgxhibOHxfozJqPtaBEAXlEKWjgR5uYw

LOGs File:
https://mclo.gs/3QIFsxs

image

**Comment by onebeastchris** [comment]
@Uniicz @jackh54 please update the debug build by re-downloading it from the links; the newly pushed commit might just resolve it ™️

File: core/src/main/java/org/geysermc/geyser/GeyserLogger.java [hunk 2]
```diff
@@ -119,6 +120,15 @@ default void debug(@Nullable Object object) {
      */
     void setDebug(boolean debug);
 
+    /**
+     * A method to debug information specific to a session.
+     */
+    default void debug(GeyserSession session, String message, Object... arguments) {
+        if (isDebug()) {
+            debug("( %s ) " + message, session.bedrockUsername(), arguments);
```
Comment (onebeastchris) [inline core/src/main/java/org/geysermc/geyser/GeyserLogger.java:128]:
```suggestion
            debug("( " + session.bedrockUsername() + " ) " + message, arguments);
```

**Comment by al1sonsz** [comment]
After the latest update, some plugin menus are not working correctly for Minecraft Bedrock players. I followed all the recommendations you provided, but they were not effective.

Now, when a menu redirects to another, it does not work, as I demonstrate in the video.

Below are the video link and the dump from Geyser.
https://dump.geysermc.org/FcCl96gvAKwihsdBp92L0xtCI7k6oiAV
https://youtu.be/XzsvDvZAxog

**Comment by onebeastchris** [comment]
@al1sonsz the dump you sent shows that you're not on the latest commit of this PR - please re-download it

**Comment by jackh54** [comment]
Latest commit seems to work for me 👍 , interacting with guis now work.

[dump](https://dump.geysermc.org/fe1c4ozyOzjMkesoSnckXa1gccHx0set)

**Comment by KazunariShimizu** [comment]
08:00:46 [情報] [Geyser-BungeeCord] Missing mapping for bedrock item BaseItemData(definition=SimpleItemDefinition(identifier=minecraft:clock, runtimeId=419, version=NONE, componentBased=false, componentData={}), damage=0, count=1, tag={
  "GeyserHash": -1706602487i,
  "display": {
    "Lore": [
      "クリックでメニューを開きます"
    ],
    "Name": "メニュー"
  }
}, canPlace=[], canBreak=[], blockingTicks=0, blockDefinition=GeyserBedrockBlock{minecraft:cyan_terracotta}, usingNetId=false, netId=0)
08:00:46 [情報] [Geyser-BungeeCord] ( tora6645 ) ClientboundOpenScreenPacket(containerId=1, type=GENERIC_9X6, title=TextComponentImpl{content="メニュー", style=StyleImpl{obfuscated=not_set, bold=not_set, strikethrough=not_set, underlined=not_set, italic=not_set, color=NamedTextColor{name="dark_gray", value="#555555"}, clickEvent=null, hoverEvent=null, insertion=null, font=null}, children=[]})
08:00:46 [情報] [Geyser-BungeeCord] ( tora6645 ) Queuing virtual inventory with id 1
08:00:46 [情報] [Geyser-BungeeCord] ( tora6645 ) Attempting to open currently delayed inventory with matching java id! 1
08:00:46 [情報] [Geyser-BungeeCord] ( tora6645 ) ContainerOpenPacket(id=1, type=CONTAINER, blockPosition=(-297, 72, -222), uniqueEntityId=-1)
08:00:47 [情報] [Geyser-BungeeCord] ( tora6645 ) ClientboundOpenScreenPacket(containerId=1, type=GENERIC_9X6, title=TextComponentImpl{content="", style=StyleImpl{obfuscated=not_set, bold=not_set, strikethrough=not_set, underlined=not_set, italic=not_set, color=null, clickEvent=null, hoverEvent=null, insertion=null, font=null}, children=[TextComponentImpl{content="メニュー", style=StyleImpl{obfuscated=false, bold=false, strikethrough=false, underlined=false, italic=false, color=NamedTextColor{name="dark_gray", value="#555555"}, clickEvent=null, hoverEvent=null, insertion=null, font=null}, children=[]}]})
08:00:47 [情報] [Geyser-BungeeCord] ( tora6645 ) Closed inventory: 1 Waiting on confirm? true
08:00:47 [情報] [Geyser-BungeeCord] ( tora6645 ) Inv (1) set pending: closing inv? true, pending inv id? -1
08:00:48 [情報] [Geyser-BungeeCord] ( tora6645 ) ContainerClosePacket(id=1, serverInitiated=false, type=NONE)
08:00:48 [情報] [Geyser-BungeeCord] ( tora6645 ) Closed inventory: 1 Waiting on confirm? false
08:00:48 [情報] [Geyser-BungeeCord] [StopSound] Stopped minecraft:block.note_block.chime -> note.chime

**Comment by KazunariShimizu** [comment]
Uploading 2025-03-31 07-46-13.mp4…

**Comment by KazunariShimizu** [comment]
https://github.com/user-attachments/assets/c60f916f-516a-4feb-a310-0607043f53af

**Comment by jackh54** [comment]
@KazunariShimizu send the dump

**Comment by KazunariShimizu** [comment]
https://dump.geysermc.org/NVwi3JeyvnYClBR2Gvumbh7T1NBezDbC

**Comment by onebeastchris** [comment]
@KazunariShimizu send a full debug log using mclo.gs, the snippet you sent isn't complete

**Comment by KazunariShimizu** [comment]
https://mclo.gs/peL1wu3

**Comment by KazunariShimizu** [comment]
https://mclo.gs/DIj46vB

**Comment by onebeastchris** [comment]
thanks for providing these, I'll add a workaround to address this tomorrow

**Comment by KazunariShimizu** [comment]
Thank you!

**Comment by Mark-Kincaid** [comment]
Latest build still doesn't work to open my CommandPanels inventory panel.

[Geyser Dump](https://dump.geysermc.org/H5z2cluZyyjpVuorrtHHj3NhWgwNl7Po)

[Server Log](https://mclo.gs/Xihiw40)

**Comment by onebeastchris** [comment]
@KazunariShimizu @Mark-Kincaid the latest commit should deal with your edge-cases as well. Please re-test the latest preview build (from the same download link)

**Comment by KazunariShimizu** [comment]
@onebeastchris onebeastchrisI tried it just now and it seems to work fine.Thank you!

**Comment by Uniicz** [comment]
Hey. Bug was fix yesterday but today seems to NOT work again.

https://mclo.gs/muYeX3j
https://dump.geysermc.org/B7a5kLLg9G3igkfkTLSJWpZDkGLTw0Vm

Video proof:
https://youtu.be/aVE2omn-OTQ

**Comment by onebeastchris** [comment]
@Uniicz update Geyser to the latest master build - this bug has been resolved there

**Comment by Uniicz** [comment]
@onebeastchris where do i find the latest master build?

**Comment by onebeastchris** [comment]
https://geysermc.org/download/

**Comment by Uniicz** [comment]
Still same problem, now even it just straight up kicks bedrock players from trying to open a GUI.

https://dump.geysermc.org/zpHw9btkQg8DPcDsBqSHNe8hZG4aSZTk
https://mclo.gs/81S03JX

**Comment by onebeastchris** [comment]
Looks like packetevents isn't updated to 1.21.5, it's throwing errors

Geyser is now emulating a 1.21.5 Java client - make sure you've updated velocity, ViaVersion, and other protocol dependant plugins to their latest versions

**Comment by Uniicz** [comment]
Everything seems to be updated to the last version, even packetevents, altought i dont see them having a 1.21.5 version. Still getting the same error.

**Comment by onebeastchris** [comment]
The log you sent shows these 3 stacktraces:
- `	[21:36:17] [Netty epoll Worker #2/WARN] [com.github.retrooper.packetevents.PacketEventsAPI]: PacketEvents caught an unhandled exception while calling your listener.`
-> either caused by PacketEvents, or a plugin using their API
- `Downstream packet error! (ClientboundMapItemDataPacket) Packet "ClientboundMapItemDataPacket" not fully read.` 
-> A warning by Geyser, but that's unrelated to inventory opening and closing 

And this is the one that's causing the kick: https://mclo.gs/81S03JX#L2896

That one is caused by PacketEvents not processing a packet correctly, and that's kicking players.

None of these are related to this PR that you're commenting on.

**Comment by FrancescoJacovo** [comment]
we also are having the same problem with everything updated to the latest version (also dev ones)

**Comment by Camotoy** [comment]
Please attach a Geyser Dump.

**Comment by FrancescoJacovo** [comment]
Here we restored the old stable packetevents because also the 2.7.1 dev gives errors https://dump.geysermc.org/MZCK2jfiXzkHV18DP3HJuQNT6RTKtSva
